spacy_experimental/span_finder/span_finder_component.py

```diff
@@ -0,0 +1,295 @@
+
+def get_span_references(docs) -> Floats2d:
+    """Create a list of reference spans for scoring"""
+    doc_spans = []
+    for doc in docs:
+        spans = set()
+        for span_key in doc.spans:
+            for span in doc.spans[span_key]:
+                spans.add((span.start, span.end))
+        doc_spans.append(spans)
+    return doc_spans
+
+
+class SpanFinder(TrainablePipe):
+    """Pipeline that learns span boundaries"""
+
+    def __init__(
+        self,
+        nlp: Language,
+        model: Model[List[Doc], Floats2d],
+        name: str = "span_finder",
+        *,
+        threshold: float = 0.5,
+        scorer: Optional[Callable],
+        candidates_key: str = "span_finder_candidates",
+    ) -> None:
+        """Initialize the span boundary detector.
+        model (thinc.api.Model): The Thinc Model powering the pipeline component.
+        name (str): The component instance name, used to add entries to the
+            losses during training.
+        threshold (float): Minimum probability to consider a prediction
+            positive.
+        scorer (Optional[Callable]): The scoring method.
+        candidates_key (str): Name of the SpanGroup the span candidates are saved to
+        """
+        self.vocab = nlp
+        self.cfg = {
+            "threshold": threshold,
+        }
+        self.model = model
+        self.name = name
+        self.scorer = scorer
+        self.candidates_key = candidates_key
+
+    def predict(self, docs: Iterable[Doc]):
+        """Apply the pipeline's model to a batch of docs, without modifying them.
+        docs (Iterable[Doc]): The documents to predict.
+        RETURNS: The models prediction for each document.
+        """
+        scores = self.model.predict(docs)
+        return scores
+
+    def set_annotations(self, docs: Iterable[Doc], scores: Floats2d) -> None:
+        """Modify a batch of Doc objects, using pre-computed scores.
+        docs (Iterable[Doc]): The documents to modify.
+        scores: The scores to set, produced by SpanFinder predict method.
+        """
+        lengths = [len(doc) for doc in docs]
+
+        offset = 0
+        scores_per_doc = []
+        for length in lengths:
+            scores_per_doc.append(scores[offset : offset + length])
+            offset += length
+
+        for doc, doc_scores in zip(docs, scores_per_doc):
+            doc.spans[self.candidates_key] = []
+            for start_token, start_token_score in zip(doc, doc_scores):
+                if start_token_score[0] > self.cfg["threshold"]:
+                    for end_token, end_token_score in zip(doc, doc_scores):
+                        if (
+                            end_token_score[1] > self.cfg["threshold"]
+                            and end_token.i + 1 > start_token.i
+                        ):
+                            doc.spans[self.candidates_key].append(
+                                doc[start_token.i : end_token.i + 1]
+                            )
```

Comment:
Although it's still a concern, my concern is less the efficiency of this particular loop than the efficiency of potentially making a very large number of very long suggestions that you probably actually don't want that you're passing to `spancat`. I think it might make sense to add a `max_span_length` parameter or something along those lines?

Other options I could imagine that are a little more complicated:

* not crossing sentence boundaries
* use the N (say, 3) nearest ends for a particular start

Obviously this depends on your use case, but I think the current approach is only suitable for very short texts.

Reply:
I've added two new parameters to the `SpanFinder`  and added more information to the README in [6422926](https://github.com/explosion/spacy-experimental/pull/10/commits/6422926e6af418467de92d2595e7d99eccd29f99) 
- `max_length`
- `min_length`
> When set to 0, the limitation is ignored.

Reply:
I would go back to a version where `starts` and `ends` were extracted initially to avoid repeatedly checking thresholds.

I think it would be faster to avoid having an extra function call to check the length constraints and you can build it shortcuts a bit more easily.

Try to order the checks so that very fast checks are run first and shortcut as much as possible:

- check whether end <= start
- check whether the limits are set
- once you have seen one end that's beyond the max_length away from the current start, you can shortcut the rest of that loop

Reply:
Thanks! I think, I was able to shorten everything quite a bit, I'm surprised that I didn't think of it earlier. Updated in [c424769](https://github.com/explosion/spacy-experimental/pull/10/commits/c4247698000669895c834bdd0f1c7a212150b7a3)